File: BCKEND/TypeScript/BancoFinanceiro/src/Model/Conta.ts
import { Agencia } from "./Agencia";
import { Cliente } from "./Cliente";
import { TipoTransacao, Transacao } from "./Transacao";

export class Conta {
    private cliente: Cliente;
    private numero: number;
    private agencia: Agencia;
    private DataAbertura: Date;
    private saldo: number;
    private status: boolean;
    private transacoes: Transacao[];

    constructor(cliente: Cliente, numero: number, agencia: Agencia) {
        this.cliente = cliente;
        this.numero = numero;
        this.agencia = agencia;
        this.DataAbertura = new Date();
        this.saldo = 0;
        this.status = true;
        this.transacoes = [];
    }

    public getCliente(): Cliente {
        return this.cliente;
    }

    public getNumero(): number {
        return this.numero;
    }

    public getAgencia(): Agencia {
        return this.agencia;
    }

    public getDataAbertura(): Date {
        return this.DataAbertura;
    }

    public getSaldo(): number {
        return this.saldo;
    }

    public getStatus(): boolean {
        return this.status;
    }

    public getTransacoes(): Transacao[] {
        return this.transacoes;
    }

    public setStatus(status: boolean): void {
        this.status = status;
    }

    // sacar
	public sacar(valor: number): boolean {
        if (this.status && valor > 0 && this.saldo >= valor) {
            this.saldo -= valor;
            this.transacoes.push(new Transacao(TipoTransacao.SAQUE, valor, "-"));
            return true;
        }else {
            console.log("Saque não autorizado.");
            return false;
        }
    }
	// depositar
	public depositar(valor: number): boolean {
        if (this.status && valor > 0) {
            this.saldo += valor;
            this.transacoes.push(new Transacao(TipoTransacao.DEPOSITO, valor, "+"));
            return true;
        }else {
            console.log("Depósito não autorizado.");
            return false;
        }
    }
	// transferir
	public transferir(valor: number, contaDestino: Conta): boolean {
        if (this.status && contaDestino.getStatus() && valor > 0 && this.saldo >= valor) {
            this.saldo -= valor;
            contaDestino.saldo += valor;
            this.transacoes.push(new Transacao(TipoTransacao.TRANSFERENCIA, valor, "-", contaDestino.getCliente()));
            contaDestino.transacoes.push(new Transacao(TipoTransacao.TRANSFERENCIA, valor, "+", this.getCliente()));
            return true;
        }else {
            console.log("Transferência não autorizada.");
            return false;
        }
    }
	// pagar
	public pagar(valor: number, infoPagamento: string): boolean {
        if (this.status && valor > 0 && this.saldo >= valor) {
            this.saldo -= valor;
            this.transacoes.push(new Transacao(TipoTransacao.PAGAMENTO, valor, "-", undefined, infoPagamento));
            return true;
        }else {
            console.log("Pagamento não autorizado.");
            return false;
        }
    }
	// extrato
    public imprimirExtrato(): void {
        console.log(`\n--- Extrato da Conta ---`);
        console.log(`Cliente: ${this.cliente.getNome()}`);
        console.log(`Número da Conta: ${this.numero}`);
        console.log(`Agência: ${this.agencia.getNumero()}`);        
        if(this.transacoes.length === 0) {
            console.log("Nenhuma transação realizada.");
        } else {
            this.transacoes.forEach((transacao, index) => {
                console.log(`${transacao.toString()}`);
            });
        }
        console.log(`Saldo Atual: R$ ${this.saldo.toFixed(2)}\n`);
    }

    // toString

}

File: BCKEND/JAVA/BancoFinanceiro/src/Model/Conta.java
package Model;

import java.util.ArrayList;
import java.util.Date;

public abstract class Conta {

	private Cliente cliente;
	private int numero;
	private Agencia ag;
	private Date data;
	private double saldo;
	private boolean status;
	private ArrayList<Transacao> transacoes;

	public Conta(Cliente cliente, int numero, Agencia ag) {
		super();
		this.cliente = cliente;
		this.numero = numero;
		this.ag = ag;
		this.data = new Date();
		this.saldo = 0.0;
		this.status = true;
		this.transacoes = new ArrayList<Transacao>();
	}

	public Conta(Cliente cliente, int numero, Agencia ag, Date data, double saldo) {
		super();
		this.cliente = cliente;
		this.numero = numero;
		this.ag = ag;
		this.data = data;
		this.saldo = saldo;
		this.status = true;
		this.transacoes = new ArrayList<Transacao>();
	}

	// sacar
	public boolean sacar(double valor) {
		if (valor > 0 ) {
			if (this.saldo >= valor) {
				this.saldo -= valor; // this.saldo = this.salado - valor
				Transacao transacao = new Transacao(TipoTransacao.Saque, valor, 
						null, null, '-');
				this.transacoes.add(transacao);
				return true;
			} else {
				System.out.println("Saldo insuficiente!");
				return false;
			}			
		} else {
			System.out.println("Valor inválido!");
			return false;
		}
	}

	// depositar
	public boolean depositar(double valor) {
		if (valor > 0) {
			this.saldo += valor;
			Transacao transacao = new Transacao(TipoTransacao.Déposito, valor,
					null, null, '+');
			this.transacoes.add(transacao);
			return true;
		} else {
			System.out.println("Valor inválido!");
			return false;
		}
	}

	// pagar
	public boolean realizarPagamento(double valor, String infoPag) {
		if (valor > 0 && infoPag != null) {
			if (this.saldo >= valor) {
				this.saldo -= valor;
				Transacao t = new Transacao(TipoTransacao.Pagamento, 
						valor, null, infoPag, '-');
				this.transacoes.add(t);
				return true;
			} else {
				System.out.println("Saldo insuficiente!");
				return false;
			}
		} else {
			System.out.println("Valores inválidos!");
			return false;
		}
	}

	// transferir
	public boolean transferir(double valor, Conta contaDestino) {
		if (valor > 0 && contaDestino != null && contaDestino.status) {
			if (this.saldo >= valor) {
				this.saldo -= valor;
				Transacao transacao = new Transacao(TipoTransacao.Transferência, valor, 
						contaDestino.getCliente(), null, '-');
				this.transacoes.add(transacao);
				contaDestino.saldo += valor;
				Transacao transacaoDestino = new Transacao(TipoTransacao.Transferência, 
						valor, this.cliente, null, '+');
				contaDestino.transacoes.add(transacaoDestino);
				return true;
			} else {
				System.out.println("Saldo insuficiente!");
				return false;
			}
		} else {
			System.out.println("Valores inválidos!");
			return false;
		}
	}

	// extrato
	public String imprimirExtrato() {
		String extrato = ".:: Extrato da Conta " + this.numero + " ::.\n";
		extrato += "Cliente: " + this.cliente.getNome() + ", CPF: " + 
				this.cliente.getCpf() + "\n" + "Agência " + this.ag.getNome() +
				", Nº " + this.ag.getNumero() + "\n";
		for (Transacao transacao : transacoes) {
			extrato += transacao + "\n";
		}
		extrato += "Saldo atual: R$ " + this.saldo;
		return extrato;
	}

	public int getNumero() {
		return numero;
	}

	public void setNumero(int numero) {
		this.numero = numero;
	}

	public Agencia getAg() {
		return ag;
	}

	public void setAg(Agencia ag) {
		this.ag = ag;
	}

	public boolean isStatus() {
		return status;
	}

	public void setStatus(boolean status) {
		this.status = status;
	}

	public Cliente getCliente() {
		return cliente;
	}

	public Date getData() {
		return data;
	}

	public double getSaldo() {
		return saldo;
	}

	public ArrayList<Transacao> getTransacoes() {
		return transacoes;
	}

}
